keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,389 @@
+If all pods represented by every `NodeReadinessGate` is ready (based on container state and readiness probes) then the node is also considered ready.
+If no pods are found with the label selector then the node is consider to be "Not Ready".
+
+The individual pods representing readiness gates are required to detect kubelet crash (or restart) quickly and reflect that in their own readiness states via probes.
+The kubelet places a special meta readiness taint on a node object if 1 or more of those pods are not ready. If the Node Lifecycle Controller observes that all the pods selected
+by `ReadinessGates` are ready, it will remove that taint immediately.
+
+If `BlocksReadiness` is set to `false`, the overall readiness of a node is not blocked by the pods under this specific readiness gate. This is useful for non-critical pods
+that should not block the readiness of a node but should be reported in the node status.
+
+### Readiness Gates from External Sources
+
+The `ReadinessGates` field can be used by external components to dictate node readiness since they can apply an arbitrary readiness gate with a reserved label selector
+that should not be used by any pods on the cluster. A common usecase here is an external cloud provider that wants to ensure no pods are scheduled on a node until
+networking for a given node is properly set up by the cloud provider. In this case, the cloud provider would apply any custom readiness gates to a node during registration
+and a separate controller would remove the readiness gate when it sees that node's networking has been properly configured. No pods matching the label selector of that
+readiness gate was required.
+
+### Reporting Node Readiness via Node Status
+
+The readiness of a node will be reported in the node's status subresource:
+
+```go
+type NodeStatus struct {
+    ...
+    Readiness []NodeReadiness `json:"readiness,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,12,rep,name=readiness"`
+}
+
+type NodeReadiness struct {
+    // The name of the node readiness gate
+    // +optional
+    Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+    // Whether a node is ready based on it's corresponding readiness gates
+    // +optional
+    Ready bool `json:"ready" protobuf:"varint,2,opt,name=ready"`
+}
+```
+
+Updates for `node.Status.Readiness` will be done by the Node Lifecycle Controller.
+
+### Limitations and Risks
+
+#### Tolerating the Readiness Taint
+
+One limitation in this design is that all system critical pods that should gate the readiness of a node must be updated to tolerate the readiness taint. Otherwise,
+those pods cannot be scheduled on a node since the readiness taint would already be applied by the kubelet on registration. If users forget to apply this toleration,
+they may accidentally gate the readiness of their nodes, preventing workloads from being scheduled there.
```

Comment:
It seems as though any pod that blocks node readiness would almost necessarily have to tolerate *all* taints... Otherwise, eg, running low on disk space would cause all the node-readiness-gating pods to be evicted, which would then cause the node to become meta-unready, which would then cause most of the other pods to be evicted (even if they explicitly indicated that they can tolerate low disk space).

Reply:
IME system critical pods tend to already tolerate all taints - but agreed this can be unexpected in many cases. If we specify readiness gate at the DaemonSet level like you suggested, we could set the right toleration when the DaemonSet is created? That doesn't sound ideal either because a DaemonSet would end up tolerating taints it shouldn't.

Reply:
> If we specify readiness gate at the DaemonSet level like you suggested, we could set the right toleration when the DaemonSet is created?

Or just require the user to create it, and reject it in validation if they didn't.

> That doesn't sound ideal either because a DaemonSet would end up tolerating taints it shouldn't.

Maybe there are use cases for a node-readiness-blocking pod to tolerate some but not all taints. Anyway, the validation check could just check that it tolerates at least the meta-ready taint, rather than that it tolerates all taints.

Reply:
Validation for tolerations sgtm